lib/ansible/modules/cloud/vmware/vmware_host_config_facts.py

```diff
@@ -0,0 +1,109 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2018, Abhijeet Kasurde <akasurde@redhat.com>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+DOCUMENTATION = r'''
+---
+module: vmware_host_config_facts
+short_description: Gathers facts about an ESXi host's advance configuration information
+description:
+- This module can be used to gather facts about an ESXi host's advance configuration information when ESXi hostname or Cluster name is given.
+version_added: '2.5'
+author:
+- Abhijeet Kasurde (@akasurde)
+notes:
+- Tested on vSphere 6.5
+requirements:
+- python >= 2.6
+- PyVmomi
+options:
+  cluster_name:
+    description:
+    - Name of the cluster from which the ESXi host belong to.
+    - If C(esxi_hostname) is not given, this parameter is required.
+  esxi_hostname:
+    description:
+    - ESXi hostname to gather facts from.
+    - If C(cluster_name) is not given, this parameter is required.
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = r'''
+- name: Gather facts about all ESXi Host in given Cluster
+  vmware_host_config_facts:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    cluster_name: cluster_name
+
+- name: Gather facts about ESXi Host
+  vmware_host_config_facts:
+    hostname: '{{ vcenter_hostname }}'
+    username: '{{ vcenter_username }}'
+    password: '{{ vcenter_password }}'
+    esxi_hostname: '{{ esxi_hostname }}'
+'''
+
+RETURN = r'''#
+'''
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils.vmware import vmware_argument_spec, PyVmomi
+
+
+class VmwareConfigFactsManager(PyVmomi):
+    def __init__(self, module):
+        super(VmwareConfigFactsManager, self).__init__(module)
+        self.cluster = self.params.get('cluster_name', None)
+        self.host = self.params.get('esxi_hostname', None)
+        self.hosts = []
+        if self.cluster:
+            self.cluster_obj = self.find_cluster_by_name(cluster_name=self.cluster)
+            if self.cluster_obj:
+                self.hosts = [host for host in self.cluster_obj.host]
+        elif self.host:
+            self.host_obj = self.find_hostsystem_by_name(host_name=self.host)
+            if self.host_obj:
+                self.hosts.append(self.host_obj)
```

Comment:
```
def __init__(self, module):
    super(VmwareConfigFactsManager, self).__init__(module)

    cluster_name = self.params.get('cluster_name', None)
    esxi_hostname = self.params.get('esxi_hostname', None)

    if cluster_name:
        cluster_obj = self.find_cluster_by_name(cluster_name=cluster_name)
        if cluster_obj:
            self.hosts = [host for host in cluster_obj.host]
        else:
            module.fail_json(changed=False, msg="Cluster '%s' not found"%cluster_name)
    elif esxi_hostname:
        esxi_host_obj = self.find_hostsystem_by_name(host_name=esxi_hostname)
        if esxi_host_obj:
            self.hosts = [ esxi_host_obj ]
        else:
            module.fail_json(changed=False, msg="ESXi '%s' not found"%esxi_hostname)
```

@Akasurde What do you think ?
No variable except `hosts` has to be attributes.
The module should fail if the cluster or ESXi isn't found.

Reply:
OK.